### PR TITLE
Query String Support

### PR DESCRIPTION
**Pull request by jhford**
Here's the initial work on supporting query strings.  This PR has the following changes:

1. Deterministic ordering for api entry insertion in the code generation logic
2. Moves all the argument processing into _processArgs which simplifies the sync and async makeApiCall functions, which means more shared code between async and sync
3. Adds support for building URLs and signed URLs for query-string params
4. Adds supprt for making api calls with query strings when _makeApiCall is called with appropriate arguments

Still to do:

1. Add tests for query-string handling
2. Change functionality of _processArgs to support query-string parameters in the call itself
3. Figure out why python 2.7 unit tests just plain don't work for me on my machine.  All development done with Python 3.6
4. Make sure that topic exchanges are also generated in a deterministic fashion 

Please ignore the changes to files other than taskcluster/client.py, taskcluster/async/asyncclient.py and genCode.py.

### CODE REVIEW

File: genCode.py [hunk 1]
```diff
@@ -175,9 +175,12 @@ def createStaticClient(name, api, genAsync=False):
 
 
     lines.append('    funcinfo = {')
-    items = functionInfo.iteritems() if six.PY2 else functionInfo.items()
-    for funcname, ref in sorted(items):
-        lines.append('        "%s": %s,' % (funcname, pprint.pformat(ref, indent=12)))
+
+    for funcname, ref in sorted(six.iteritems(functionInfo)):
```
Comment (jonasfj) [inline genCode.py]: This looks sane, but would have been nice if auto-generated code changes were in a separate commit :)

Comment (jhford) [reply genCode.py]: yah.... sorry about that.

**Comment by gregarndt** [comment]
@escapewindow Hey Aki, mind checking this out? This is the implementation that you should be use for your wrapper script that you might write.

**Comment by jhford** [comment]
@jonasfj I don't know how to redo the review request, so I'll ping you and try to figure it out.  Here's an example of how these changes will work
```python
jhford-macbook:~/taskcluster/taskcluster-client.py $ ./.tox/py35/bin/python test-cont-tok.py
Response 0 gave us 1000 more tasks
Response 1 gave us 1000 more tasks
Response 2 gave us 1000 more tasks
Response 3 gave us 1000 more tasks
Task Group JzTGxwxhQ76_Tt1dxkaG5g has 3982 tasks
jhford-macbook:~/taskcluster/taskcluster-client.py $ cat test-cont-tok.py
import taskcluster
queue = taskcluster.Queue()
i = 0
tasks = 0
outcome = queue.listTaskGroup('JzTGxwxhQ76_Tt1dxkaG5g')
while outcome.get('continuationToken'):
    print('Response %d gave us %d more tasks' % (i, len(outcome['tasks'])))
    if outcome.get('continuationToken'):
        outcome = queue.listTaskGroup('JzTGxwxhQ76_Tt1dxkaG5g', query={'continuationToken': outcome.get('continuationToken')})
    i += 1
    tasks += len(outcome.get('tasks', []))
print('Task Group %s has %d tasks' % (outcome['taskGroupId'], tasks))

```

A couple notes:

1. We're still supporting the two old versions of the calling convention, through some checks.  Please let me know if there are any edge cases you think the checks wouldn't cover.  `test/test_client.py:TestProcessArgs` is the relevant test case, `client.py:BaseClient._processArgs()` is the relevant function
2. I'd like to deprecate the `method(payload, k1=v1)` style of calling methods but I'm not totally convinced whether or not to do it.  Basically, this format is only allowed when all the kwargs are exactly the same as the api args, but doesn't support query strings, so it should be safe.
3. I was thinking of supporting a `paginationHandler` function which can be passed in as a kwarg using the `method(query={}, payload={})` form which gets called on each response.
4. I reverted all the code-generation stuff, we'll need to do an api update before releasing
5. I won't push to pypi for a little while to let people test, since semver isn't a python standard

**Comment by jhford** [comment]
@srfraser do you think you could try out this branch with your tools?

File: taskcluster/client.py [hunk 2]
```diff
@@ -219,62 +222,125 @@ def genBewit():
...
         return self._makeHttpRequest(entry['method'], route, payload)
 
-    def _processArgs(self, entry, *args, **kwargs):
-        """ Take the list of required arguments, positional arguments
-        and keyword arguments and return a dictionary which maps the
-        value of the given arguments to the required parameters.
-
-        Keyword arguments will overwrite positional arguments.
+    def _processArgs(self, entry, *_args, **_kwargs):
+        """ Figure out, given an entry, positional and keyword arguments, what
```
Comment (imbstack) [inline taskcluster/client.py]: Tiny Nit: I think this reads a bit better as "Given an entry, positional and keyword arguments, figure out what the query-string options, payload and api arguments are."

File: taskcluster/client.py [hunk 3]
```diff
@@ -219,62 +222,125 @@ def genBewit():
...
+        the query-string options, payload and api arguments are.
         """
 
+        # We need the args to be a list so we can mutate them
+        args = list(_args)
+        kwargs = copy.deepcopy(_kwargs)
+        log.debug(args)
```
Comment (imbstack) [inline taskcluster/client.py]: If these are being left in, perhaps they should be something like  "${entry} called with args ${args} and kwargs ${kwargs}". Not that it has to be exactly that, but it might be a bit nicer.

It is also possible that I don't remember how python logging works and this is just fine as it is.

File: taskcluster/client.py [hunk 4]
```diff
@@ -219,62 +222,125 @@ def genBewit():
...
         reqArgs = entry['args']
-        data = {}
+        routeParams = {}
+
+        query = {}
+        payload = None
+        kwApiArgs = {}
+
+        # There are three formats for calling methods:
+        #   1. method(v1, v1, payload)
```
Comment (jonasfj) [inline taskcluster/client.py]: Do this mode not support query params?

Note: I ask because this is the mode we use in all other client libraries.
Typically, it looks like: `service.method(v1, v2, payload, query)` where `query` is optional, and `payload` is required if a payload is accepted by the method.

Comment (jhford) [reply taskcluster/client.py]: it doesn't, and it cannot.  In other languages that's simple because you know how many position arguments are required, and for sure if payload is expected, so you can make the last arg optional.  In the python client, this is complicated by supportingn bare kwargs and args in the same call.

I think it's even better to force `query={}` because it just makes it so much clearer what's going on.

**Comment by jhford** [comment]
Not sure when the commit is going to show up here, github's cache invalidation is busted right now, but I pushed a change to this branch that supports:
```
jhford-macbook:~/taskcluster/taskcluster-client.py $ cat pagination-example.py
import taskcluster
queue = taskcluster.Queue()

responses = []

def x(y):
    print("%d tasks fetched" % len(y.get('tasks', [])))
    responses.append(y)

queue.listTaskGroup('JzTGxwxhQ76_Tt1dxkaG5g', paginationHandler=x)

tasks = 0
for response in responses:
    tasks += len(response.get('tasks', []))

print("%d requests fetch %d tasks" % (len(responses), tasks))

jhford-macbook:~/taskcluster/taskcluster-client.py $ ./.tox/py35/bin/python pagination-example.py
1000 tasks fetched
1000 tasks fetched
1000 tasks fetched
1000 tasks fetched
982 tasks fetched
5 requests fetch 4982 tasks
```

File: taskcluster/client.py (hunk 2 above)
Comment (jhford) [reply taskcluster/client.py]: Agreed

File: taskcluster/client.py (hunk 3 above)
Comment (jhford) [reply taskcluster/client.py]: they're not supposed to be left in, but I couldn't for the life of me remember where i'd typed them.  i'll remove them

**Comment by escapewindow** [comment]
This adds the paginationHandler to asyncclient:

```diff
diff --git a/taskcluster/async/asyncclient.py b/taskcluster/async/asyncclient.py
index 57362d3..82af6dd 100644
--- a/taskcluster/async/asyncclient.py
+++ b/taskcluster/async/asyncclient.py
@@ -4,7 +4,6 @@ from __future__ import absolute_import, division, print_function

 import os
 import logging
-import copy
 import hashlib
 import hmac
 import datetime
@@ -60,11 +59,29 @@ class AsyncBaseClient(BaseClient):
         """ This function is used to dispatch calls to other functions
         for a given API Reference entry"""

-        routeParams, payload, query, _, _ = self._processArgs(entry, *args, **kwargs)
+        x = self._processArgs(entry, *args, **kwargs)
+        routeParams, payload, query, paginationHandler, paginationLimit = x
         route = self._subArgsInRoute(entry, routeParams)
-        log.debug('Route is: %s', route)

-        return await self._makeHttpRequest(entry['method'], route, payload)
+        # TODO: Check for limit being in the Query of the api ref
+        if paginationLimit and 'limit' in entry.get('query', []):
+            query['limit'] = paginationLimit
+
+        if query:
+            _route = route + '?' + urllib.parse.urlencode(query)
+        else:
+            _route = route
+        response = await self._makeHttpRequest(entry['method'], _route, payload)
+
+        if paginationHandler:
+            paginationHandler(response)
+            while response.get('continuationToken'):
+                query['continuationToken'] = response['continuationToken']
+                _route = route + '?' + urllib.parse.urlencode(query)
+                response = await self._makeHttpRequest(entry['method'], _route, payload)
+                paginationHandler(response)
+        else:
+            return response

     async def _makeHttpRequest(self, method, route, payload):
         """ Make an HTTP Request for the API endpoint.  This method wraps
```

This is a test script that uses it:

```python
#!/usr/bin/env python
import asyncio
import pprint
from taskcluster.async import Queue


def main():
    task_group = 'JzTGxwxhQ76_Tt1dxkaG5g'
    output = []
    queue = Queue()

    def handler(contents):
        tasks = contents['tasks']
        print("handler called! adding {} tasks".format(len(tasks)))
        output.extend(contents['tasks'])

    loop = asyncio.get_event_loop()
    loop.run_until_complete(
        queue.listTaskGroup('JzTGxwxhQ76_Tt1dxkaG5g', paginationHandler=handler)
    )
    print('Task Group %s has %d tasks' % (task_group, len(output)))
    pprint.pprint(output)


__name__ == '__main__' and main()
```

**Comment by jhford** [comment]
Merged to master, thanks everyone!
